### PR TITLE
Subcompositor fixes

### PR DESCRIPTION
**Pull request by wmww**
Fixed subsurface offset issue and sync behavior. Input coming soon!

### CODE REVIEW

File: src/server/frontend_wayland/wl_surface.cpp [hunk 1]
```diff
@@ -66,6 +79,13 @@ mf::WlSurface::~WlSurface()
     *destroyed = true;
     if (auto session = get_session(client))
         session->destroy_buffer_stream(stream_id);
+    // it is possible to destroy a wl_surface before destroying its role (at least if the role is a subsurface)
+    // the role needs to be told its surface is invalid. Also children subsurfaces need to be unmapped.
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/wl_surface.cpp]: Is this comment a TODO?

If so, it should include the token "TODO", if not it isn't clear.

Comment (wmww) [reply src/server/frontend_wayland/wl_surface.cpp]: This comment is not correct. Roles are destroyed when the surface is destroyed. Will edit commit to remove.

Comment (wmww) [reply src/server/frontend_wayland/wl_surface.cpp]: Turns out there was a nugget of truth to that comment. I resolved the issue in 
1ae9429

File: src/server/frontend_wayland/wl_surface.cpp [hunk 2]
```diff
@@ -66,6 +79,12 @@ mf::WlSurface::~WlSurface()
     *destroyed = true;
     if (auto session = get_session(client))
         session->destroy_buffer_stream(stream_id);
+    role->destroy();
+}
+
+bool mf::WlSurface::synchronized() const
```
Comment (RAOF) [inline src/server/frontend_wayland/wl_surface.cpp]: This is the WlSurfaceRole abstraction getting leaky :grin:

We probably don't want WlSurface to implement the all methods that all wl_surface roles allow. I don't have a better idea at the moment; this is just a thought to look out for.

File: src/server/frontend_wayland/wl_surface.cpp [hunk 3]
```diff
@@ -35,6 +35,19 @@
...
+{
+    if (source.buffer)
+        buffer = source.buffer;
+
+    if (source.buffer_offset)
+        buffer_offset = source.buffer_offset;
+
+    frame_callbacks->insert(frame_callbacks->end(),
+                            source.frame_callbacks->begin(),
+                            source.frame_callbacks->end());
```
Comment (AlanGriffiths) [inline src/server/frontend_wayland/wl_surface.cpp]: That should be:

```
    frame_callbacks.insert(end(frame_callbacks), begin(source.frame_callbacks), end(source.frame_callbacks));
```

**Pull request merged**